istioctl/pkg/rbac/upgrade.go

```diff
@@ -0,0 +1,296 @@
+func (ug *Upgrader) upgradeLocalFile(fileName string) (string, error) {
+	err := ug.createRoleAndBindingLists(fileName)
+	if err != nil {
+		return "", err
+	}
+	newFile, newFilePath, err := checkAndCreateNewFile(fileName)
+	if err != nil {
+		return "", err
+	}
+	for _, serviceRole := range ug.serviceRoles {
+		if err := ug.updateRoleToWorkloadLabels(serviceRole); err != nil {
+			return "", err
+		}
+		role := ug.upgradeServiceRole(serviceRole)
+		err = writeToFile(newFile, role)
+		if err != nil {
+			return "", err
+		}
+	}
+	for _, serviceRoleBinding := range ug.serviceRoleBindings {
+		authzPolicy := ug.createAuthorizationPolicyFromRoleBinding(serviceRoleBinding)
+		err = writeToFile(newFile, authzPolicy)
+		if err != nil {
+			return "", err
+		}
+	}
+	fmt.Println(fmt.Sprintf("New authorization policies have been upgraded to %s", newFilePath))
+	return newFilePath, nil
+}
+
+// writeToFile writes data from `config` to `file`.
+func writeToFile(file *os.File, config model.Config) error {
+	var stringBuilder strings.Builder
+	schema, exists := configDescriptor.GetByType(config.Type)
+	if !exists {
+		return fmt.Errorf("unknown kind %q for %v", crd.ResourceName(config.Type), config.Name)
+	}
+	obj, err := crd.ConvertConfig(schema, config)
+	if err != nil {
+		return fmt.Errorf("could not decode %v: %v", config.Name, err)
+	}
+	stringBuilder.WriteString(fmt.Sprintf("%s%s/%s\n", apiVersionField, config.Group, config.Version))
+	stringBuilder.WriteString(fmt.Sprintf("%s%s\n", kindField, obj.GetObjectKind().GroupVersionKind().Kind))
+	stringBuilder.WriteString(fmt.Sprintf("%s\n", metadataField))
+	stringBuilder.WriteString(fmt.Sprintf("  %s%s\n", nameField, config.Name))
+	stringBuilder.WriteString(fmt.Sprintf("  %s%s\n", namespaceField, config.Namespace))
+	stringBuilder.WriteString(fmt.Sprintf("%s\n", specField))
+	spec, err := yaml.Marshal(obj.GetSpec())
+	if err != nil {
+		return err
+	}
+	specLines := strings.Split(string(spec), "\n")
+	for i, line := range specLines {
+		// If specLines has an empty line at the end, ignore it for better formatting.
+		if i == len(specLines)-1 && line == "" {
+			break
+		}
+		stringBuilder.WriteString(fmt.Sprintf("  %s\n", line))
+	}
+	stringBuilder.WriteString("---\n")
+	if _, err := file.WriteString(stringBuilder.String()); err != nil {
+		return fmt.Errorf("failed to write %s to %s because %v", config.Name, file.Name(), err)
+	}
+	return nil
+}
+
+// createRoleAndBindingLists creates lists of model.Configs to store ServiceRole and ServiceRoleBinding policies.
+func (ug *Upgrader) createRoleAndBindingLists(fileName string) error {
+	rbacFileBuf, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		return fmt.Errorf("failed to read file %s", fileName)
+	}
+	configsFromFile, _, err := crd.ParseInputs(string(rbacFileBuf))
+	if err != nil {
+		return err
+	}
+	for _, config := range configsFromFile {
+		if config.Type == configDescriptor.Types()[0] {
+			// ServiceRole
```

Comment:
Unnecessary comment, here and other places.

Reply:
Will do. Thanks